### PR TITLE
Assert that REST params are consumed iff supported

### LINKED ISSUE #113854: Forbid consumption of unsupported query params
https://github.com/elastic/elasticsearch/pull/112478 accidentally dropped support for the `?timeout` param, reinstated in https://github.com/elastic/elasticsearch/pull/113852. I think we could have detected this sooner: it should be an error for `prepareRequest` to attempt to consume a parameter that isn't in the set returned from `supportedQueryParameters()`, regardless of whether this parameter is supplied in the request or not.

### PR DESCRIPTION
**Pull request by DaveCTurner**
REST APIs which declare their supported parameters must consume exactly
those parameters: consuming an unsupported parameter means that requests
including that parameter will be rejected, whereas failing to consume a
supported parameter means that this parameter has no effect and should
be removed.

This commit adds an assertion to verify that we are consuming the
correct parameters.

Closes #113854

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java [hunk 1]
```diff
@@ -147,6 +150,19 @@ public void close() {
         }
     }
 
+    private boolean assertConsumesSupportedParams(@Nullable Set<String> supported, RestRequest request) {
+        if (supported != null) {
+            final var supportedAndCommon = new TreeSet<>(supported);
+            supportedAndCommon.addAll(List.of("error_trace", "filter_path", "format", "pretty", "human"));
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:156]: Do we have these common params defined anywhere so this isn't an arbitrary list?

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java]: Eh not all of them. There's `org.elasticsearch.rest.BaseRestHandler#ALWAYS_SUPPORTED` but that doesn't include `error_trace`. I pushed a commit that uses `ALWAYS_SUPPORTED`. Not a huge deal here in any case, if we get this wrong then this assertion will trip.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
Bah this had a bug, I reverted it and will open a fresh PR.

**Pull request merged**